<a href="https://colab.research.google.com/github/mudesir/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/Mudesir_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

In [1]:
### From the raw data we see that the dataset doesn't have column headers and have some missed values lets 
# fix the header and the cell values assigned by "?" clean our data
import pandas as pd
column_headers = ['party', 'handicapped-infants', 'water-project-cost-sharing', 'adoption-of-the-budget-resolution', 
                    'physician-fee-freeze', 'el-salvador-aid', 'religious-groups-in-schools', 'anti-satellite-test-ban', 
                    'aid-to-nicaraguan-contras', 'mx-missile', 'immigration', 'synfuels-corporation-cutback', 'education-spending', 
                    'superfund-right-to-sue', 'crime', 'duty-free-exports', 'export-administration-act-south-africa']
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data', names=column_headers)
df.head()                                                                                                                          

party  ... export-administration-act-south-africa
0  republican  ...                                      y
1  republican  ...                                      ?
2    democrat  ...                                      n
3    democrat  ...                                      y
4    democrat  ...                                      y

[5 rows x 17 columns]

In [2]:
import numpy as np

df = df.replace({'y': 1, 'n': 0, '?': np.NaN})
print(df.shape)
df.head()

(435, 17)


party  ...  export-administration-act-south-africa
0  republican  ...                                     1.0
1  republican  ...                                     NaN
2    democrat  ...                                     0.0
3    democrat  ...                                     1.0
4    democrat  ...                                     1.0

[5 rows x 17 columns]

In [0]:
df.isnull().sum()

Party                                       0
handicapped-infants                        12
water-project-cost-sharing                 48
adoption-of-the-budget-resolution          11
physician-fee-freeze                       11
el-salvador-aid                            15
religious-groups-in-schools                11
anti-satellite-test-ban                    14
aid-to-nicaraguan-contras                  15
mx-missile                                 22
immigration                                 7
synfuels-corporation-cutback               21
education-spending                         31
superfund-right-to-sue                     25
crime                                      17
duty-free-exports                          28
export-administration-act-south-africa    104
dtype: int64

In [0]:
df['Party'].value_counts()

democrat      267
republican    168
Name: Party, dtype: int64

In [0]:
#
dem = df[df['Party']=='democrat']
rep = df[df['Party']=='republican']

In [0]:
dem.head()

Party  ...  export-administration-act-south-africa
2  democrat  ...                                     0.0
3  democrat  ...                                     1.0
4  democrat  ...                                     1.0
5  democrat  ...                                     1.0
6  democrat  ...                                     1.0

[5 rows x 17 columns]

In [0]:
rep.head()


Party  ...  export-administration-act-south-africa
0   republican  ...                                     1.0
1   republican  ...                                     NaN
7   republican  ...                                     1.0
8   republican  ...                                     1.0
10  republican  ...                                     0.0

[5 rows x 17 columns]

 2 sample t-test on Bills vote democrats support more than republicans 
1. Null Hypothhesis: democrats support the same proportin as republican on Bills
$H_0: \bar{x}_d = \bar{x}_r$
2. Alternative Hypothesis: democrats support are not equal with republican on issues
$H_0: \bar{x}_d  !=  \bar{x}_r$
3. Confidence level: 99%

The Bills are: ***'handicapped-infants', 'water-project-cost-sharing', 'adoption-of-the-budget-resolution', 'physician-fee-freeze', 'el-salvador-aid', 'religious-groups-in-schools', 'anti-satellite-test-ban', 'aid-to-nicaraguan-contras', 'mx-missile', 'immigration',           'synfuels-corporation-cutback', 'education-spending', 'superfund-right-to-sue', 'crime', 'duty-free-exports', 'export-administration-act-south-africa'***

In [0]:
#handicapped-infants
from scipy import stats
stats.ttest_ind(dem['handicapped-infants'], rep['handicapped-infants'],nan_policy='omit')

Ttest_indResult(statistic=9.205264294809222, pvalue=1.613440327937243e-18)

**Conclusion**: Given a t-statistic of 9.2, and p-value of 0, we **reject** the null hypothesis that democrats suport for **handicapped-infants** bill is equal to repabilican. This suggests the alternative that the democrats do not support the bill at the same bill. And also from the sign of t-statiscs the Democrat support the bill **more than** the republicans.

In [0]:
# water-project-cost-sharing
stats.ttest_ind(dem['water-project-cost-sharing'], rep['water-project-cost-sharing'],nan_policy='omit')

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

**Conclusion**: Given a t-statistic of 0.089, and p-value of 0.929, we **fail to reject** the null hypothesis that democrats suport for **water-project-cost-sharing** bill is equal to repabilican. And also the Democrat support the bill **not more than** the republicans.

In [0]:
# adoption-of-the-budget-resolution
stats.ttest_ind(dem['adoption-of-the-budget-resolution'], rep['adoption-of-the-budget-resolution'], nan_policy='omit')

Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)

**Conclusion**: Given a t-statistic of 23.213, and p-value of 0, we **reject** the null hypothesis that democrats suport for **adoption-of-the-budget-resolution** bill is equal to repabilican. This suggests the alternative that the democrats do not support the bill at the same bill. And also from the sign of t-statiscs the Democrat support the bill **more than** the republicans.

In [0]:
# physician-fee-freeze
stats.ttest_ind(dem['physician-fee-freeze'], rep['physician-fee-freeze'],nan_policy='omit')

Ttest_indResult(statistic=-49.36708157301406, pvalue=1.994262314074344e-177)

In [0]:
# el-salvador-aid
stats.ttest_ind(dem['el-salvador-aid'], rep['el-salvador-aid'],nan_policy='omit')

Ttest_indResult(statistic=-21.13669261173219, pvalue=5.600520111729011e-68)

**Conclusion**: Given a t-statistic of -21.137, and p-value of 0, we **reject** the null hypothesis that democrats suport for **el-salvador-aid** bill is equal to repabilican. This suggests the alternative that the democrats do not support the bill at the same bill. And also from the sign of t-statistic the Democrat support the bill **less than** the republicans.

In [0]:
# religious-groups-in-schools
stats.ttest_ind(dem['religious-groups-in-schools'], rep['religious-groups-in-schools'],nan_policy='omit')

Ttest_indResult(statistic=-9.737575825219457, pvalue=2.3936722520597287e-20)

**Conclusion**: Given a t-statistic of -9.738, and p-value of 0, we **reject** the null hypothesis that democrats suport for **religious-groups-in-schools** bill is equal to repabilican. This suggests the alternative that the democrats do not support the bill at the same bill. And also from the sign of t-statistic the Democrat support the bill **less than** the republicans.

In [0]:
# anti-satellite-test-ban
stats.ttest_ind(dem['anti-satellite-test-ban'], rep['anti-satellite-test-ban'],nan_policy='omit')

Ttest_indResult(statistic=12.526187929077842, pvalue=8.521033017443867e-31)

**Conclusion**: Given a t-statistic of 12.526, and p-value of 0, we **reject** the null hypothesis that democrats suport for **anti-satellite-test** bill is equal to repabilican. This suggests the alternative that the democrats do not support the bill at the same bill. And also from the sign of t-statistic the Democrat support the bill **more than** the republicans.

In [0]:
# aid-to-nicaraguan-contras
stats.ttest_ind(dem['aid-to-nicaraguan-contras'], rep['aid-to-nicaraguan-contras'],nan_policy='omit')

Ttest_indResult(statistic=18.052093200819733, pvalue=2.82471841372357e-54)

**Conclusion**: Given a t-statistic of 18.052, and p-value of 0, we **reject** the null hypothesis that democrats suport for **aid-to-nicaraguan-cotras** bill is equal to repabilican. This suggests the alternative that the democrats do not support the bill at the same bill. And also from the sign of t-statistic the Democrat support the bill **more than** the republicans.

In [0]:
# mx-missile
stats.ttest_ind(dem['mx-missile'], rep['mx-missile'],nan_policy='omit')

Ttest_indResult(statistic=16.437503268542994, pvalue=5.03079265310811e-47)

**Conclusion**: Given a t-statistic of 16.438, and p-value of 0, we **reject** the null hypothesis that democrats suport for **mx-missile** bill is equal to repabilican. This suggests the alternative that the democrats do not support the bill at the same bill. And also from the sign of t-statistic the Democrat support the bill **more than** the republicans.

In [0]:
# immigration
stats.ttest_ind(dem['immigration'], rep['immigration'],nan_policy='omit')

Ttest_indResult(statistic=-1.7359117329695164, pvalue=0.08330248490425066)

**Conclusion**: Given a t-statistic of -1.736, and p-value of 0.083, we **fail to reject** the null hypothesis that democrats suport for **immigration** bill is equal to repabilican. And also from the sign of t-statistic the Democrat support the bill **less than** the republicans.

In [0]:
# synfuels-corporation-cutback
stats.ttest_ind(dem['synfuels-corporation-cutback'], rep['synfuels-corporation-cutback'],nan_policy='omit')

Ttest_indResult(statistic=8.293603989407588, pvalue=1.5759322301054064e-15)

**Conclusion**: Given a t-statistic of 8.294, and p-value of 0, we **reject** the null hypothesis that democrats suport for **synfuels-corporation-cutback** bill is equal to repabilican. This suggests the alternative that the democrats do not support the bill at the same bill. And also from the sign of t-statistic the Democrat support the bill **more than** the republicans.

In [0]:
# education-spending
stats.ttest_ind(dem['education-spending'], rep['education-spending'],nan_policy='omit')

Ttest_indResult(statistic=-20.500685724563073, pvalue=1.8834203990450192e-64)

**Conclusion**: Given a t-statistic of -20.501, and p-value of 0, we **reject** the null hypothesis that democrats suport for **education-spending** bill is equal to repabilican. This suggests the alternative that the democrats do not support the bill at the same bill. And also from the sign of t-statistic the Democrat support the bill **less than** the republicans.

In [0]:
# superfund-right-to-sue
stats.ttest_ind(dem['superfund-right-to-sue'], rep['superfund-right-to-sue'],nan_policy='omit')

Ttest_indResult(statistic=-13.51064251060933, pvalue=1.2278581709672758e-34)

**Conclusion**: Given a t-statistic of -13.511, and p-value of 0, we **reject** the null hypothesis that democrats suport for **superfund-right-to-sue** bill is equal to repabilican. This suggests the alternative that the democrats do not support the bill at the same bill. And also from the sign of t-statistic the Democrat support the bill **less than** the republicans.

In [0]:
# crime
stats.ttest_ind(dem['crime'], rep['crime'],nan_policy='omit')

Ttest_indResult(statistic=-16.342085656197696, pvalue=9.952342705606092e-47)

**Conclusion**: Given a t-statistic of -16.342, and p-value of 0, we **reject** the null hypothesis that democrats suport for **crime** bill is equal to repabilican. This suggests the alternative that the democrats do not support the bill at the same bill. And also from the sign of t-statistic the Democrat support the bill **less than** the republicans.

In [0]:
# duty-free-exports
stats.ttest_ind(dem['duty-free-exports'], rep['duty-free-exports'],nan_policy='omit')

Ttest_indResult(statistic=12.853146132542978, pvalue=5.997697174347365e-32)

**Conclusion**: Given a t-statistic of 12.853, and p-value of 0, we **reject** the null hypothesis that democrats suport for **duty-free-exports** bill is equal to repabilican. This suggests the alternative that the democrats do not support the bill at the same bill. And also from the sign of t-statistic the Democrat support the bill **more than** the republicans.

In [0]:
# export-administration-act-south-africa
stats.ttest_ind(dem['export-administration-act-south-africa'], rep['export-administration-act-south-africa'],nan_policy='omit')

Ttest_indResult(statistic=6.849454815841208, pvalue=3.652674361672226e-11)

**Conclusion**: Given a t-statistic of 6.849, and p-value of 0, we **reject** the null hypothesis that democrats suport for **export-adminstration-act-south-africa** bill is equal to repabilican. This suggests the alternative that the democrats do not support the bill at the same bill. And also from the sign of t-statistic the Democrat support the bill **more than** the republicans.

## Stretch Goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then check your Answers using Scipy!